### ISSUE TITLE
Support filter context (state bag) value templates

### ISSUE THREAD

**Issue by AlexanderYastrebov** [zalando/skipper#1444]
**Is your feature request related to a problem? Please describe.**

It is often needed to transfer/reuse values within route. Currently special-purpose filters are created, e.g.:
- `queryToHeader`, `headerToQuery`
- `setContextRequestHeader`
- `setDynamicBackendHostFromHeader`

**Describe the solution you would like**

* Support template resolution in filter values from context (state bag).
* Add *getter* filter variants to store value into context, e.g. `getRequestHeader("X-Foo-Header", "foo")`
* Add context *setter*, e.g. `set("first", "value")` or `set("second", "from ${first}")`. Might also be `setOnRequest`+`setOnResponse` pair.
* Optional/future: Add context-only filters to work with values stored in context, e.g. `replace("myVar", "^zalando\.(\w+)$", "www.zalando.$1")`

Examples:
* `setContextRequestHeader("X-Uid", "auth-user")` becomes `setRequestHeader("X-Uid", "${auth-user}")`
* `queryToHeader("foo-query-param", "X-Foo-Header")` could be `getQuery("foo-query-param", "myParam") -> setResponseHeader("X-Foo-Header", "${myParam}")`
* `queryToHeader("access_token", "Authorization", "Bearer %s")` (format string variant) could be `getQuery("access_token", "tok") -> setResponseHeader("Authorization", "Bearer ${tok}")`
* `headerToQuery("X-Foo-Header", "foo-query-param")` could be `getRequestHeader("X-Foo-Header", "hh") -> setQuery("foo-query-param", "${hh}")`
* `cookieToHeader("jwt", "Authorization", "Bearer ")` (#1006) would be `getCookie("jwt", "token") -> setResponseHeader("Authorization", "Bearer ${token}")`
* `setDynamicBackendHostFromHeader("X-Forwarded-Host") -> <dynamic>;` (#908) could be `getRequestHeader("X-Forwarded-Host", "backend:dynamic:host") -> <dynamic>;`
* `setDynamicBackendHost("example.com") -> <dynamic>;` could be `set("backend:dynamic:host", "example.com") -> <dynamic>;`
* Check token against path/header usecase (https://github.com/zalando/skipper/issues/804#issuecomment-427301435):
`
Path("/foo/:bar") -> oauthTokenintrospectionAllKV("url", "key", "${bar}"); * -> getRequestHeader("foo", "checkedHeader") -> oauthTokenintrospectionAllKV("url", "key", "${checkedHeader}")`

**Describe alternatives you've considered (optional)**

Current approach introduces special-purpose filters, e.g. #1384, #1006, #908, #804 

**Additional context (optional)**

For some filters (e.g. `setRequestHeader("X-Uid", "${auth-user}")`) proposed change could be breaking as it would interpret previously literal value as template. It is expected to be a rare case for existing route definitions as filter needs to use template marker (`${...}`) and value must be present in context for final result to be different from literal.

Template support already exists and is used by `modPath` and `modQuery`.

**Would you like to work on it?**

At first I would like to understand whether proposal makes sense.

**Comment by AlexanderYastrebov** [comment]
#1453 could be implemented given:
- `Path` predicate stores path params to state bag
- `setResponseHeader` filter supports value template substitution

**Comment by AlexanderYastrebov** [comment]
Related to https://github.com/zalando/skipper/issues/1367

**Comment by AlexanderYastrebov** [comment]
https://github.com/zalando/skipper/pull/1462#discussion_r448948138 demonstrates how header/query/cookie values can be projected for template variable resolution thus eliminating the need for *getter* filter variants to store value into context:
```
// queryToHeader("foo-query-param", "X-Foo-Header")
setRequestHeader("X-Foo-Header", "${request.query.foo-query-param}")

// queryToHeader("access_token", "Authorization", "Bearer %s")
setRequestHeader("Authorization", "Bearer ${request.query.access_token}")

// headerToQuery("X-Foo-Header", "foo-query-param")
setQuery("foo-query-param", "{request.header.X-Foo-Header}")

// cookieToHeader("jwt", "Authorization", "Bearer ") 
setResponseHeader("Authorization", "Bearer ${request.cookie.jwt}")
```

**Comment by szuecs** [comment]
@AlexanderYastrebov I think the idea of getters that store things in the context for modification and that can be referenced makes a lot of sense and seems to be really powerful!

**Comment by AlexanderYastrebov** [comment]
@szuecs To me _[projection](https://github.com/zalando/skipper/issues/1444#issuecomment-653167980)_ idea that came later seems more powerful. It does not require persisting intermediate values into context - they could be used right away in many cases. And to store value (if really needed) it would not require N getters (for headers, query, path param, etc) but rather single filter, e.g. `setContext`:
```
setContext("foo", "${response.header.X-Foo}")
setContext("bar", "${request.query.bar}")
```

**Comment by szuecs** [comment]
@AlexanderYastrebov sgtm, so can you express the final semantics of it?

Operations I could imagine:
- get `"${response.header.X-Foo}"`
- set `"${response.header.X-Foo}=${request.query.bar}"`
- modify `TBD`

I guess `setContext("foo", "${response.header.X-Foo}")` will store the value `Response.Header.Get("X-Foo")` in the projectionContext (type maybe `map[string]interface{}` or `map[string]string`) called "foo" (`projectionContext["foo"] = Response.Header.Get("X-Foo")`. Does this makes sense?

So we could have `modifyContext("foo", <what-do-we-here>)`???

**Comment by AlexanderYastrebov** [comment]
@szuecs Let me summarize what we have at the moment and possible extensions to it.

### Current state
1. Template variable resolution for header filter family (e.g. `setRequestHeader`) implemented by #1462 
- [documented](https://github.com/zalando/skipper/blob/master/docs/reference/filters.md#setrequestheader)
- supports path params and state bag values
2. Template variable resolution for `setQuery` filter
- undocumented, marked as [experimental](https://github.com/zalando/skipper/blob/071da49db445a8fc3060f45e269ea3cd390e9501/filters/builtin/query.go#L40-L44)
- supports only path params
3. Template variable resolution for `setPath` filter
- undocumented, marked as [experimental](https://github.com/zalando/skipper/blob/071da49db445a8fc3060f45e269ea3cd390e9501/filters/builtin/path.go#L33-L37) as well
- supports only path params

Next logical step would be to use the same (unify) template variable resolution function between `set*Header`, `setQuery` and `setPath` filters, update documentation and remove experimental mark.

### Extension
Once the same template resolution function is used by all filters it may be further extended to support additional variable lookup sources beside path params and state bag. 
The [_projection_](https://github.com/zalando/skipper/issues/1444#issuecomment-653167980) idea suggests to use prefix to denote variable lookup source so e.g. variables with prefix `request.query` are looked up in request query params. For example
```
setRequestHeader("X-Foo-Header", "${request.query.foo-query-param}")
```
sets request header `X-Foo-Header` to the value of request query param `foo-query-param` which is equivalent to  `queryToHeader("foo-query-param", "X-Foo-Header")` filter usage.

Additional feasible variable lookup sources are:
- request headers (via `request.header` prefix)
- response headers (via `response.header` prefix)
- request query (via `request.query` prefix)
- cookies (via `request.cookie` prefix)

With extended template resolution function, all filters that support variable resolution (i.e.  `set*Header`, `setQuery` and `setPath`) would support it and therefore it would be possible to copy data from any of header, query and path to any of header query and path. 

Moreover template string supports multiple variables so it is possible to combine multiple values into one. E.g. set header to concatenation of a strings, query param and another header:
```
setResponseHeader("X-Greeting", "Hello ${request.query.username} from ${response.header.Server}")
```

Also _concatenate modification_ of values becomes trivial and does not require new filters, e.g.:
```
setResponseHeader("Server", "My ${response.header.server}")
```
appends `My ` prefix to the `Server` response header. 

### Filter context aka State Bag

Filter context and State bag are synonyms according e.g. to [`setContextRequestHeader`](https://github.com/zalando/skipper/blob/master/docs/reference/filters.md#setcontextrequestheader) filter docs:
> Set headers for requests using values from the filter context (state bag).

In my [comment](https://github.com/zalando/skipper/issues/1444#issuecomment-700566455) I have used `setContext` imaginary filter to actually store value into state bag, i.e. I did not meant any new context (in fact we would need a pair of such filters one for request and one for response execution paths). But I think with extension described previously there is no practical need to store values into state bag because all the values are readily accessible as template variables.

**Comment by szuecs** [comment]
@AlexanderYastrebov thanks for the summary, it helps me a lot to understand.
This seems to be a great advance and I only have one question: 
Do we want to make it possible to interact with statebag, for example we could peek into validated and parsed JWT tokens to select specific values?
If so we could use `${stateBag["foo"]}` to access it. It can of course be postponed to later.

**Comment by AlexanderYastrebov** [comment]
> Do we want to make it possible to interact with statebag

@szuecs Header filter family resolve variables from statebag already as implemented by #1462 
https://github.com/zalando/skipper/blob/071da49db445a8fc3060f45e269ea3cd390e9501/filters/builtin/headerfilter.go#L257-L268

so one can use `setRequestHeader("X-Uid", "${auth-user}")` which is equivalent to `setContextRequestHeader("X-Uid", "auth-user")`.

If we unify template resolution function based on this implementation then other filters (`setQuery` and `setPath`) would support it as well.

If we would need to _store_ values into statebag then a pair of new filters will do it but I would not rush with implementing this until real use case appears.

**Comment by AlexanderYastrebov** [comment]
The common facility to resolve template placeholders was introduced and utilized by `(set|append)(Request|Response)Header`, `setPath` and `setQuery` filters, other filters can be enhanced to use it as well. Several sources for placeholders were implemented, see https://github.com/zalando/skipper/blob/master/docs/reference/filters.md#template-placeholders
